core/src/main/java/org/apache/calcite/rel/rules/FilterCorrelateRule.java

```diff
@@ -86,8 +86,10 @@ public void onMatch(RelOptRuleCall call) {
         aboveFilters,
         JoinRelType.INNER,
         false,
-        !corr.getJoinType().toJoinType().generatesNullsOnLeft(),
-        !corr.getJoinType().toJoinType().generatesNullsOnRight(),
+        corr.getJoinType().returnsJustFirstInput()
+            || !corr.getJoinType().toJoinType().generatesNullsOnLeft(),
+        !corr.getJoinType().returnsJustFirstInput()
```

Comment:
Can't this be replaced with `corr.getJoinType()==SemiJoinType.INNER`. The only time where we can push the filter into the right relation is for SemiJoinType.INNER. In any other the case the right side either generates nulls either does not exist.